packages/gatsby/cache-dir/dev-loader.js

```diff
@@ -138,6 +143,41 @@ class DevLoader extends BaseLoader {
       ___emitter.emit(`pageQueryResult`, newPageData)
     }
   }
+
+  handleDirtyPageQueryMessage(msg) {
+    msg.payload.dirtyQueries.forEach(dirtyQueryId => {
+      if (dirtyQueryId === `/dev-404-page/` || dirtyQueryId === `/404.html`) {
+        // those pages are not on demand so skipping
+        return
+      }
+
+      const normalizedId = normalizePagePath(dirtyQueryId)
+
+      // We can't just delete items in caches, because then
+      // using history.back() would show dev-404 page
+      // due to our special handling of it in root.js (loader.isPageNotFound check)
+      // so instead we mark it as stale and instruct loader's async methods
+      // to refetch resources if they are marked as stale
+
```

Comment:
This was a pretty useful comment for me 👍

Reply:
It took me some time to figure out, so figured I would share my findings for anyone wondering why it's done this way